{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Nucleic acids 1: Genome Assembly"
     ]
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Introduction"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "You have each been given a set of 100bp reads, generated by an Illumina Genome Analyzer IIx sequencer.\n",
      "\n",
      "The objective of this practical is to assemble these reads into longer contigs ready for gene annotation.\n",
      "\n",
      "You are encouraged to edit text, add additional cells and save the state of this workbook to keep a record of what you have done in this practical.\n",
      "\n",
      "\n",
      "The nucleic acids practicals will be assessed by a short final report due *1pm Mon 2nd Feb*.\n",
      "\n",
      "This report will itself be in the form of an ipython notebook, more details to follow."
     ]
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "FASTQ files"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Download your own dataset from http://www.theosysbio.bio.ic.ac.uk/pinney/nucleic1 and save it in the same directory as this notebook.\n",
      "\n",
      "Start a new linux terminal, navigate to the same directory and take a look at this [FASTQ](http://en.wikipedia.org/wiki/FASTQ_format) format file using the linux **`less`** command. \n",
      "\n",
      "What information do the different lines contain?\n",
      "\n",
      "How many lines does the file contain (use the **`wc`** command in linux).\n",
      "\n",
      "How many reads do you have?\n"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The genome assembler that we will use is [**Phrap**](http://en.wikipedia.org/wiki/Phrap). \n",
      "\n",
      "Unfortunately this software does not accept FASTQ files but requires that the information is separated into a DNA sequence (FASTA) and a quality score (QUAL) file. \n",
      "\n",
      "Preparing these files is your first programming task."
     ]
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Reading FASTQ using PyCogent"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The [**PyCogent**](http://pycogent.org/) Python module provides a uniform interface to many commonly used bioinformatics resources and programs.\n",
      "\n",
      "We will use PyCogent throughout these practicals to make it easier for you to concentrate on the science without getting held up by too many technical details."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The following code uses a [parser](http://en.wikipedia.org/wiki/Parsing#Parser) from PyCogent to read a FASTQ file. \n",
      "\n",
      "It prints out the label, sequence and quality lines for each record, with blank lines between records.\n",
      "\n",
      "Change the filename to make it load your own FASTQ file, then run the code within the notebook using *CTRL-Enter*."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "#from cogent.parse.fastq import MinimalFastqParser\n",
      "#for label, seq, qual in MinimalFastqParser('Chan_Jonathan_single.fastq',strict=False):\n",
      "#    print label\n",
      "#    print seq\n",
      "#    print qual\n",
      "#    print"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 24
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Output a FASTA file"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The same code is copied below. Edit this code to print out a FASTA-format file containing the DNA sequences."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "import cogent\n",
      "\n",
      "from cogent.parse.fastq import MinimalFastqParser\n",
      "\n",
      "out_file = open(\"DNA_Sequence.fna\", \"w\") \n",
      "\n",
      "for label, seq, qual in MinimalFastqParser('Chan_Jonathan_single.fastq',strict=False):   \n",
      "    out_file.write('>' + label + '\\n') \n",
      "    out_file.write(seq + '\\n\\n')\n",
      "    \n",
      "out_file.close()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "IOError",
       "evalue": "[Errno 2] No such file or directory: 'Chan_Jonathan_single.fastq'",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[1;31mIOError\u001b[0m                                   Traceback (most recent call last)",
        "\u001b[1;32m<ipython-input-1-552ba34d51cb>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[0mout_file\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"DNA_Sequence.fna\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"w\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 7\u001b[1;33m \u001b[1;32mfor\u001b[0m \u001b[0mlabel\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mseq\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mqual\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mMinimalFastqParser\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'Chan_Jonathan_single.fastq'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mstrict\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      8\u001b[0m     \u001b[0mout_file\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'>'\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mlabel\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m'\\n'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      9\u001b[0m     \u001b[0mout_file\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mseq\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m'\\n\\n'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
        "\u001b[1;32m/bioinformatics/epd-7.3-1-rh5-x86_64/lib/python2.7/site-packages/cogent/parse/fastq.pyc\u001b[0m in \u001b[0;36mMinimalFastqParser\u001b[1;34m(data, strict)\u001b[0m\n\u001b[0;32m     15\u001b[0m     \"\"\"\n\u001b[0;32m     16\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[0mtype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mstr\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 17\u001b[1;33m         \u001b[0mdata\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     18\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     19\u001b[0m     \u001b[1;31m# fastq format is very simple, defined by blocks of 4 lines\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
        "\u001b[1;31mIOError\u001b[0m: [Errno 2] No such file or directory: 'Chan_Jonathan_single.fastq'"
       ]
      }
     ],
     "prompt_number": 1
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Once you have it working, copy your code into a new script (using emacs) and save it as `fastq_2_dna.py`\n",
      "\n",
      "Run your program and redirect the output to a new file to produce your FASTA file \n",
      "(use the file extension `'.fna'` for FASTA nucleic acid)."
     ]
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Output a QUAL file"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The QUAL file format has header lines just like FASTA, followed by numerical quality scores separated by spaces.\n",
      "\n",
      "FASTQ represents the quality score for each basepair using [ASCII](http://en.wikipedia.org/wiki/ASCII) characters. It is simple to convert these characters back to a numerical score, e.g. for the FASTQ quality score \"h\":\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "   print ord('h')-64"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "40\n"
       ]
      }
     ],
     "prompt_number": 23
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "\n",
      "The number 64 is the 'offset' used by Illumina (i.e. the number added to the score to produce the ASCII code).\n",
      "\n",
      "We need to convert each qual string from the FASTQ file into a list of numbers. In Python it is easy to iterate over each character of a string using a FOR loop:\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "from cogent.parse.fastq import MinimalFastqParser\n",
      "\n",
      "out_file = open(\"DNA_Sequence.qual\", \"w\")\n",
      "\n",
      "for label, seq, qual in MinimalFastqParser('Chan_Jonathan_single.fastq', strict=False):\n",
      "    out_file.write('>' + label + '\\n')   # add a line here to print the FASTA-style header line\n",
      "    for x in qual:\n",
      "        num = ord(x)-64\n",
      "        out_file.write(str(num) + ' ') # add a line within this loop to print the numerical quality score as a number, followed by a space but no carriage return.\n",
      "    out_file.write('\\n\\n')# this command just prints a carriage return after all of the scores.\n",
      "\n",
      "out_file.close()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 21
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Once you have this working, copy your code into a new file and save it as `fastq_2_qual.py`\n",
      "\n",
      "As before, run your program and redirect to a new output file.\n",
      "Phrap requires that this file has the same name as your FASTA file with the additional extension `'.qual'`\n"
     ]
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Sequence assembly with Phrap"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Once you have extracted the sequences and quality scores into separate files, you can assemble your reads using the linux command\n",
      "````\n",
      "phrap XXXX.fna\n",
      "````\n",
      "\n",
      "\n"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Phrap will load the sequences and quality scores, assemble the reads and produce several output files with different extensions:\n",
      "\n",
      "**`.contigs`** contains the contigs that phrap has managed to build from the sequence fragments.\n",
      "\n",
      "**`.contigs.qual`** contains the quality scores assigned by phrap for each base in the sequence.\n",
      "\n",
      "**`.singlets`** contains any sequences that phrap was unable to assemble into a contig.\n",
      "\n",
      "**`.log`** is a log file of the assembly process.\n",
      "\n",
      "**`.problems`** and **`.problems.qual`** contain data for any sequences for which there was a problem.\n",
      "\n"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The only file you need is `.contigs`, as this contains the assembly itself. \n",
      "\n",
      "Change the extension to `.contigs.fasta` (use the **`mv`** command in linux) so that PyCogent will recognise this as a FASTA file."
     ]
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Drawing a histogram"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The following Python code will produce a simple histogram using the [matplotlib](http://matplotlib.org/) module.\n",
      "\n",
      "Try running it here."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "%matplotlib inline\n",
      "import matplotlib.pyplot as plt\n",
      "\n",
      "x = [3.2,6.4,4.6,4.5,5.4,5.8,2.9,6.6,7.4]\n",
      "\n",
      "plt.hist(x)\n",
      "plt.xlabel('value')\n",
      "plt.ylabel('frequency')\n",
      "plt.ylim(0,10)\n",
      "plt.xlim(0,10)\n",
      "\n",
      "plt.show()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "display_data",
       "png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAEKCAYAAAAb7IIBAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFCtJREFUeJzt3X9Q1HXix/HXGimEoJSjdBmnjSaCFoSIZ9EtDcOZZmjE\nKH/kD2wm6S41b7yZ85wTZ8wbxy7PaRIbO5zUUUqbphJ/pFNrWgbaLyegyJJT6SwLVFCWQdnvH35d\n40yXXXY/n8X38zHDDLvy2fdrF/bz8vP+/FiHx+PxCABgnB52BwAA2IMCAABDUQAAYCgKAAAMRQEA\ngKEoAAAwVMgKoLCwUAMGDNDIkSO99zU1NSk3N1cJCQmaNGmSmpubQzU8AMCHkBXAzJkztXPnzg73\nlZSUKCEhQd98840GDhyoNWvWhGp4AIAPISuAzMxMxcXFdbivsrJSs2bNUq9evVRYWKiKiopQDQ8A\n8MHSfQAHDx5UYmKiJCkxMVGVlZVWDg8A+IUIKwfr7FUnHA5HiJMAwI3Jn6v7WLoFkJ6erpqaGklS\nTU2N0tPTr/mzHo+HL49Hixcvtj1DuHzxWvBa8Fpc/8tflhZARkaGSktL1dLSotLSUo0ZM8bK4QEA\nvxCyAigoKNDYsWNVW1urO++8U+vWrVNRUZGOHTumYcOGqb6+XrNnzw7V8AAAH0K2D2Dz5s2/ev9b\nb70VqiFvSE6n0+4IYYPX4gpeiyt4LQLn8AQycRRiDocjoPksADCZv+tOLgUBAIaiAADAUBQAABiK\nAgAAQ1EAAGAoCgAADEUBAIChKAAAMBQFAACGogAAwFAUAAAYigIAAENRAABgKAoAAAxFAQCAoSgA\nADAUBQAAhqIAAMBQFAAAGIoCAABDUQAAYCgKAAAMRQEAgKEoAAAwFAUAAIaiAADAUBQAABiKAgAA\nQ1EAAGAoCgAADEUBAIChKAAAMBQFAACGogAAwFC2FMDatWs1duxYpaWlad68eXZEAADjWV4ADQ0N\nWrZsmXbv3q2DBw+qtrZWu3btsjoGABgvwuoBo6Ki5PF4dObMGUnS+fPnFRcXZ3UMADCe5VsAUVFR\nKikp0aBBgxQfH6/7779fo0ePtjoGABjP8i2AU6dOqaioSNXV1YqLi1N+fr7Ky8s1YcKEDj9XXFzs\n/d7pdMrpdFobFADCnMvlksvlCnh5h8fj8QQvjm/l5eXasGGDysrKJEklJSWqq6vT8uXLr4RyOGRx\nLADo9vxdd1o+BZSZmalDhw6poaFBra2t2rFjh3JycqyOAQDGs3wKKDY2VosWLdLkyZN1/vx5jRs3\nTllZWVbHAADjWT4F1BlMAQGA/8J+CggAEB4oAAAwFAUAAIaiAADAUBQAABiKAgAAQ1EAAGAoCgAA\nDEUBAIChKAAAMBQFAACGogAAwFAUAAAYigIAAENRAABgKAoAAAxFAQCAoSgAADAUBQAAhqIAAMBQ\nFAAAGIoCAABDUQAAYCgKAAAMRQEAgKEoAAAwFAUAAIaiAADAUBQAABiKAgAAQ1EAAGAoCgAADEUB\nAIChKAAAMBQFAACGogAAwFC2FMC5c+c0ffp03X333UpKStLHH39sRwwAMFqEHYMuXrxYCQkJevnl\nlxUREaFz587ZEQMAjObweDweqwdNSUnRgQMHFBUV9av/7nA4ZEMsAOjW/F13+pwCSktL00svvaTG\nxsYuBbvsxIkTcrvdKioqUkZGhpYvXy632x2UxwYAdJ7PKaCysjKtW7dO6enpGjVqlGbOnKmcnBw5\nHI6ABnS73aqtrdWKFSuUnZ2tp556Sq+//rqmTZvW4eeKi4u93zudTjmdzoDGA4AblcvlksvlCnj5\nTk8Btbe3a9u2bSoqKlKPHj00e/ZszZ8//5rTONczfPhw1dTUSJJ27Nih9evXa/PmzVdCMQUEAH4L\n+hSQJH3xxReaP3++FixYoLy8PG3ZskUOh0M5OTkBhRw6dKgqKirU3t6u8vJyZWdnB/Q4AIDA+dwC\nSEtLU58+ffTkk08qLy9PvXr18v7b5MmT9eabb/o9aG1traZNmya3263s7GwtWbJE0dHRV0KxBQAA\nfvN33emzAL777jvdddddXQ7mDwoAAPwX9CmgV155RadPn/bebmxs1KJFiwJLBwAIGz4LoLy8XH37\n9vXejouL07Zt20IaCgAQej4LoH///vr++++9t+vr6xUXFxfSUACA0PN5HsCsWbM0fvx4TZ06VR6P\nR2VlZVq4cKEV2QAAIdSp8wDq6uq0ZcsWSdLjjz+uwYMHhzYUO4EBwG9BPwrIDhQAAPgv6EcB7dmz\nRw899JD69u2rmJgYxcTEKDY2tkshAQD287kFMGrUKK1atUq/+93v1KOHNR8fwBYAAPgv6FsAPXv2\nVFpammUrfwCANXweBZSZmalJkyYpPz/fez6Aw+HQY489FvJwAIDQ8VkAP/zwg+Lj47V///4O91MA\nANC9cRQQANwggr4P4OjRoyoqKlJqaqok6fDhw1q6dGngCQEAYcFnARQXF2vixIne2yNHjuzw4S0A\ngO7JZwHU1tZq/Pjx3tvt7e3q2bNnSEMBAELP507gBx54QJ988okkqbW1VSUlJfrDH/4Q8mAAgNDy\nuQUwb948rV69WidPntRdd92lqqoqzZkzx4psAIAQ6vRRQBcuXLBs+oejgADAf/6uO31OAS1ZssT7\noA6Hw3v/3//+98ASAgDCgs8poOjoaEVHR6t3795yu9167bXX9MMPP1iRDQAQQn6fCNbQ0KDc3Fzt\n27cvVJmYAgKAAAT9RLD/1dLSorNnz/q7GAAgzPjcBzBy5Ejv962trWpvb+dMYAC4AficAqqrq/N+\nHxkZqfj4+FBnYgoIAAIQ9I+EbGhouO4D3HrrrZ0erLMoAADwX9ALYNCgQTp27Jh69eol6dI0UEJC\nghwOhxwOh7777ruuJf61UBQAAPgt6DuBH330Ua1du1aNjY1qbGzUv//9b02cOFFHjx4NycofAGAN\nn1sAiYmJqq6u9n4kZHt7u5KSkvTVV1+FLhRbAADgt6BPAf35z39WW1ubZsyYIY/How0bNuimm27S\nP//5zy6HvWYoCgAA/Bb0Ajh9+rRKS0u1c+dOSdLDDz+swsJC9enTp2tJrxeKAgAAvwW9AC5raWlR\nVFRUwMH8QQEAgP+CvhP4888/14QJE5SUlOS9/fTTTweeEAAQFnwWwHPPPafly5erb9++kqSUlBTt\n3bs35MEAAKHlswC+//57jRgxwnu7tbVVt9xyS0hDAQBCz+e1gHJycvTWW29Jko4dO6YXX3xRubm5\nIQ8GAAgtn1sAc+fO1WeffaaLFy/q4YcfVt++ffXMM89YkQ0AEELXPQrowoULmjFjhjZu3BjUQS9e\nvKhRo0Zp4MCBeuedd64OxVFAAOC3oB4FFBERoaNHj+rUqVNdDvZLq1atUlJSUoePmAQAWMvnPoAR\nI0YoMzNTjzzyiG6//XZJl1pm/vz5AQ144sQJbd++XX/729/0wgsvBPQYAICuu+YWwBNPPCFJeu21\n11RQUKDevXurublZzc3NampqCnjAZ599VitWrPBeWwgAYI9rbgF8/fXX+s9//qOEhAQ988wzQZmT\n37Ztm/r376/U1FS5XK7r/mxxcbH3e6fTKafT2eXxAeBG4nK5fK5Lr+eaO4E3btyolStX6ssvv9Rv\nfvObjgsF+DkACxcu1IYNGxQRESG3262zZ88qLy9P69evv+rx2QkMAP4J+rWAZs+erTVr1nQ52P/a\nu3evnn/+eY4CAoAgCfq1gEKx8r+Mo4AAwD6dvhqoldgCAAD/BX0LAABwY6IAAMBQFAAAGIoCAABD\nUQAAYCgKAAAMRQEAgKEoAAAwFAUAAIaiAADAUBQAABiKAgAAQ1EAAGAoCgAADEUBAIChKAAAMBQF\nAACGogAAwFAUAAAYigIAAENRAABgKAoAAAxFAQCAoSgAADAUBQAAhqIAAMBQFAAAGIoCAABDUQAA\nYCgKAAAMRQEAgKEoAAAwFAUAAIaiAADAUBQAABjK8gI4fvy4srKylJycLKfTqU2bNlkdAQAgyeHx\neDxWDnjy5EmdPHlSKSkp+umnnzR69Gh98cUXiomJuRLK4ZDFsQCg2/N33Wn5FkB8fLxSUlIkSf36\n9VNycrIOHTpkdQwAMJ6t+wCOHDmiqqoqjR492s4YAGCkCLsGbmpq0pQpU7Ry5UpFR0df9e/FxcXe\n751Op5xOp3XhoNjYW9XU1BjQsjExcTp7tiHIiXzrjpmBrnC5XHK5XAEvb/k+AElqa2vThAkTNH78\neM2bN+/qUOwDsJ3D4ZAU6O/Ant9fd8wMBJO/607LC8Dj8Wj69Onq16+fXnjhhV8PRQHYrjuuTLtj\nZiCYwr4A9u/frwcffFD33HPP/79hpX/84x8aN27clVAUgO2648q0O2YGginsC6AzKAD7dceVaXfM\nDART2B8GCgAIDxQAABiKAgAAQ1EAAGAoCgAADEUBAIChKAAAMBQFAACGogAAwFAUAAAYigIAAENR\nAABgKAoAAAxFAQCAoSgAADAUBQAAhqIAAMBQEXYHALq72Nhb1dTU2IVHuFlSmzHLxsTE6ezZhgDH\nRTDxkZD4Vd3x4xXtyty1cS+N3ZXc3XFZ3t+hwUdCAgA6hQIAAENRAABgKAoAAAxFAQCAoSgAADAU\nBQAAhqIAAMBQFAAAGIoCAABDUQAAYCgKAAAMRQEAgKEoAAAwFAUAAIaiAADAUBQAABjKlgL44IMP\nNHz4cA0dOlQvvviiHRG6DZfLZXeEsMFr8UsuuwOEDf4uAmdLAcydO1cvv/yy9uzZo5deekk//fST\nHTG6Bf64r+C1+CWX3QHCBn8XgbO8AM6cOSNJevDBB/Xb3/5WOTk5qqiosDoGABjP8gI4ePCgEhMT\nvbeTkpL08ccfWx0DAIwXYXeAa3E4HHZHCBtLliyxaeTAfweh+v35fi3sytzV5xvI8pdfC7tyh8/f\nh33vke7N8gJIT0/XggULvLerqqo0bty4Dj/j8XisjgUAxrF8CqhPnz6SLh0JVFdXp927dysjI8Pq\nGABgPFumgP71r3/pqaeeUltbm+bMmaN+/frZEQMAjGbLYaC///3vVVNToyNHjmjOnDne+zk/4JLj\nx48rKytLycnJcjqd2rRpk92RbHfx4kWlpqZq4sSJdkex1blz5zR9+nTdfffdxh9AsXbtWo0dO1Zp\naWmaN2+e3XEsVVhYqAEDBmjkyJHe+5qampSbm6uEhARNmjRJzc3NPh8nrM4E5vyAS26++WatXLlS\nVVVV2rp1qxYtWqSmpia7Y9lq1apVSkpKMv7ggMWLFyshIUGHDx/W4cOHNXz4cLsj2aKhoUHLli3T\n7t27dfDgQdXW1mrXrl12x7LMzJkztXPnzg73lZSUKCEhQd98840GDhyoNWvW+HycsCkAzg+4Ij4+\nXikpKZKkfv36KTk5WYcOHbI5lX1OnDih7du368knnzT+AIE9e/Zo4cKFioyMVEREhHefmmmioqLk\n8Xh05swZtbS06Pz584qLi7M7lmUyMzOver6VlZWaNWuWevXqpcLCwk6tP8OmADg/4NcdOXJEVVVV\nGj16tN1RbPPss89qxYoV6tEjbP5cbXHixAm53W4VFRUpIyNDy5cvl9vttjuWLaKiolRSUqJBgwYp\nPj5e999/v9HvEanjOjQxMVGVlZU+lzH7HRXmmpqaNGXKFK1cuVLR0dF2x7HFtm3b1L9/f6Wmphr/\nv3+3263a2lrl5eXJ5XKpqqpKr7/+ut2xbHHq1CkVFRWpurpadXV1OnDggMrLy+2OZatA3h9hUwDp\n6en66quvvLerqqo0ZswYGxPZq62tTXl5eXriiSeUm5trdxzbfPTRR3r77bc1ePBgFRQU6L333tO0\nadPsjmWLIUOGaNiwYZo4caKioqJUUFCgHTt22B3LFpWVlRozZoyGDBmi2267Tfn5+frggw/sjmWr\n9PR01dTUSJJqamqUnp7uc5mwKQDOD7jC4/Fo1qxZGjFihHFHN/yvZcuW6fjx4zp69KjKysr00EMP\naf369XbHss3QoUNVUVGh9vZ2lZeXKzs72+5ItsjMzNShQ4fU0NCg1tZW7dixQzk5OXbHslVGRoZK\nS0vV0tKi0tLSTv0HOmwKQLpyfkB2draefvppY88P+PDDD7Vx40a99957Sk1NVWpq6lV7/E1l+lFA\nzz//vObOnav77rtPkZGRmjp1qt2RbBEbG6tFixZp8uTJeuCBB3TvvfcqKyvL7liWKSgo0NixY1Vb\nW6s777xT69atU1FRkY4dO6Zhw4apvr5es2fP9vk4Do/pE6sAYKiw2gIAAFiHAgAAQ1EAAGAoCgAA\nDEUBAJ3Uu3dvuyMAQUUBAJ1k+iGouPFQADDWX//6V61evdp7u7i4WM8995yys7N13333afz48dq7\nd+9Vy7lcrg6Xpf7Tn/6kV199VZL09ddfe6/V88c//lE///xz6J8IECAKAMaaMmVKh2vpbNmyRTNm\nzNCbb76pTz/9VGvWrFFxcbHPx3E4HN6tgwULFmjhwoWqqKhQcnKyXnnllVDFB7osbD8UHgi1lJQU\n/fjjj/rvf/+rH3/8UXFxcbrjjju0dOlSbd++XefOndO3336rM2fOdOqyy6dOndK+ffv06KOPSrr0\nITaDBg0K8bMAAkcBwGj5+fnaunWrTp48qalTp8rlcmnfvn3atWuXoqOj1b9//6sKIDIyUq2trd7b\nl6d5Ll68qNtuu02fffaZ5c8DCARTQDDalClTtHnzZm3dulX5+fmqr6/XHXfcoZiYGJWVlamhoeGq\nZVJSUlRdXa3m5mbV19fr3XfflXTpg3wGDx6sN954Qx6PR21tbaqurrb6KQGdRgHAaElJSWpubtbA\ngQM1YMAATZo0SadPn9bw4cO1f/9+JSUleX/28jx/ZGSk/vKXv2jMmDEqLCzscBXK1atX6/3331dK\nSopSU1N14MABy58T0FlcDA4ADMUWAAAYigIAAENRAABgKAoAAAxFAQCAoSgAADDU/wGclkHOq3G9\nVwAAAABJRU5ErkJggg==\n",
       "text": [
        "<matplotlib.figure.Figure at 0x209a090>"
       ]
      }
     ],
     "prompt_number": 25
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The first line,\n",
      "````python\n",
      "%matplotlib inline\n",
      "````\n",
      "is a bit of iPython \"magic\" to let you use matplotlib within this workbook. For a stand-alone Python script, just leave this line out."
     ]
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "A histogram of contig lengths"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Adapt the code below to load your set of contigs and make a histogram of their lengths.\n",
      "\n",
      "**Hint:** the command \n",
      "````python\n",
      "x.append('apple')\n",
      "```` \n",
      "adds the string `'apple'` to the list `x`.\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "%matplotlib inline\n",
      "import matplotlib.pyplot as plt\n",
      "from cogent import LoadSeqs, DNA\n",
      "\n",
      "# load a set of contigs from the file XXXX.fasta as a SequenceCollection object and store in the variable 'seqs'.\n",
      "seqs = LoadSeqs('DNA_Sequence.fna.contigs.fasta', moltype=DNA, format='fasta', aligned=False)\n",
      "\n",
      "# make an empty list to contain contig lengths\n",
      "lengths = []\n",
      "\n",
      "# iterate over all sequences in 'seqs'\n",
      "\n",
      "for seqid, seq in seqs.items():\n",
      "    lengths.append(len(seq)) ## add a line within this loop to append the sequence length to 'lengths'\n",
      "\n",
      "print lengths\n",
      "## add code below to draw a histogram of the data stored in 'lengths'\n",
      "\n",
      "plt.hist(lengths)\n",
      "plt.xlabel('length')\n",
      "plt.ylabel('frequency')\n",
      "plt.xlim(0, 4000)\n",
      "plt.ylim(0, 10)\n",
      "plt.show()\n",
      "\n",
      "lengths \n",
      "out_file = open(\"longest.fasta\", \"w\")\n",
      "out_file.write('>' + label + '\\n')\n",
      "out_file.write(str(seq))\n",
      "out_file.close()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "TypeError",
       "evalue": "unsupported operand type(s) for +: 'int' and 'str'",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
        "\u001b[1;32m<ipython-input-3-121527e8199f>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     12\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     13\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mseqid\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mseq\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mseqs\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 14\u001b[1;33m     \u001b[0mlengths\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mseq\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mseqid\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m## add a line within this loop to append the sequence length to 'lengths'\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     15\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     16\u001b[0m \u001b[1;32mprint\u001b[0m \u001b[0mlengths\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
        "\u001b[1;31mTypeError\u001b[0m: unsupported operand type(s) for +: 'int' and 'str'"
       ]
      }
     ],
     "prompt_number": 3
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "How long is your longest contig?\n",
      "\n",
      "Can you write a script to print out only the longest contig in FASTA format?\n",
      "\n",
      "Write this contig to a new file called `longest.fasta` - we will work with this in the next practical."
     ]
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Optional exercise"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Using the code you have already written as a starting point, write a single Python script that will read in your FASTQ file, create the FASTA and QUAL files, run Phrap and produce a report on the assembled contigs. \n",
      "\n",
      "(You will need to make use of Python's [file handling commands](http://www.pythonforbeginners.com/systems-programming/reading-and-writing-files-in-python/) and the [subprocess.call](https://docs.python.org/2/library/subprocess.html#subprocess.call) function for running external Linux commands).\n",
      "\n",
      "How can you make your script usable on any FASTQ file without having to edit the code?"
     ]
    }
   ],
   "metadata": {}
  }
 ]
}